core/src/rust/src/ingestion.rs

```diff
@@ -0,0 +1,266 @@
+            writer.run([UserOperation::Delete(delete_term), UserOperation::Add(doc)])?;
+
+            handle.changes_pending.store(true, Ordering::SeqCst);
+        } else {
+            writer.add_document(doc)?;
+        }
+
+        handle.changes_pending.store(true, Ordering::SeqCst);
+
+        Ok(())
+    });
+}
+
+fn prepare_tantivy_doc(
+    handle: &IndexHandle,
+    ingesting_doc: &mut IngestingDocument,
+) -> JavaResult<TantivyDocument> {
+    let mut map_values = std::mem::take(&mut ingesting_doc.map_values);
+
+    // Insert map columns we've built up
+    for (key, value) in map_values.drain() {
+        ingesting_doc
+            .doc
+            .add_object(handle.schema.get_field(&key)?, value);
+    }
+
+    // Build final facet for field list
+    let mut field_names = std::mem::take(&mut ingesting_doc.field_names);
+    field_names.sort();
+
+    for field in field_names {
+        add_facet(
+            handle,
+            ingesting_doc,
+            field_constants::LABEL_LIST,
+            &[field.as_str()],
+        )?;
+    }
+
+    let doc = std::mem::take(&mut ingesting_doc.doc);
+
+    Ok(doc)
+}
+
+fn add_facet(
+    handle: &IndexHandle,
+    ingesting_doc: &mut IngestingDocument,
+    name: &str,
+    value: &[&str],
+) -> JavaResult<()> {
+    if !name.is_empty() && !value.is_empty() {
+        ingesting_doc.doc.add_facet(
+            handle.schema.get_field(&facet_field_name(name))?,
+            Facet::from_path(value),
+        );
+    }
+
+    Ok(())
+}
+
+#[no_mangle]
+pub extern "system" fn Java_filodb_core_memstore_TantivyNativeMethods_00024_removePartKeys(
+    mut env: JNIEnv,
+    _class: JClass,
+    handle: jlong,
+    keys: JIntArray,
+) {
+    jni_exec(&mut env, |env| {
+        let handle = IndexHandle::get_ref_from_handle(handle);
+        let mut terms = vec![];
+
+        let field = handle.schema.get_field(field_constants::PART_ID)?;
+
+        let len = env.get_array_length(&keys)?;
+        let mut part_ids = vec![0i32; len as usize];
+
+        env.get_int_array_region(&keys, 0, &mut part_ids)?;
+
+        for part_id in part_ids {
+            terms.push(Term::from_field_i64(field, part_id as i64));
+        }
+
+        let query = Box::new(TermSetQuery::new(terms));
+
+        let writer = handle.writer.read()?;
+        writer.delete_query(query)?;
+
+        handle.changes_pending.store(true, Ordering::SeqCst);
+
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "system" fn Java_filodb_core_memstore_TantivyNativeMethods_00024_removePartitionsEndedBefore(
+    mut env: JNIEnv,
+    _class: JClass,
+    handle: jlong,
+    ended_before: jlong,
+    return_deleted_count: jboolean,
+) -> jint {
+    jni_exec(&mut env, |_| {
+        let handle = IndexHandle::get_ref_from_handle(handle);
+
+        let query = RangeQuery::new_i64_bounds(
+            field_constants::END_TIME.to_string(),
+            Bound::Included(0),
+            Bound::Included(ended_before),
```

Comment:
Should it be `Excluded`?

Reply:
This matches the Lucene behavior of LongRange queries, where both bounds are included.